Lazor/py project (#58)
NotANameServer/Not-a-Hub: langages/python/comment-structurer-un-projet-python.md

Comment structurer un projet Python

> ===================================

Reviewer: ```suggestion
# Comment structurer un projet Python
```
je préfèrerais que l'on évite de mélanger les styles `#` `=` et `-`

* [Introduction](#introduction)
* [Mise en place du squelette du projet](#mise-en-place-du-squelette-du-projet)
* [Adaptation d'un programme existant](#adaptation-dun-programme-existant)
* [Isolation dans un environnement virtuel](#isolation-dans-un-environnement-virtuel)
* [Ajout de tests automatiques](#ajout-de-tests-automatiques)
* [Création d'une archive installable](#création-dune-archive-installable)
* [Configuration de VSCode](#configuration-de-vscode)
* [Configuration de PyCharm](#configuration-de-pycharm)
* [Choix des technologies](#choix-des-technologies)
* [Références](#références)

Introduction

> ------------

Reviewer: ```suggestion
## Introduction
```

La plupart des cours que nous recommandons[^1] [^2] [^3] pour apprendre Python se concentrent sur l'enseignement de la programmation *au travers* de Python et se réservent bien d'enseigner comment *structurer et déployer* une application ou une bibliothèque écrite en Python.

> Il existe en ligne une pléthore de tutoriels[^4] [^5] [^6] sur comment structurer son projet et comment le déployer. Cependant l'éco-système de Python a tellement évolué depuis 2018 sur tout ce qui est relatif au *packaging* a qu'il a rendu bon nombre de ces tutoriels caduque[^7] [^8].

Reviewer 2: Le dernier "a" est en trop :

> Python a tellement évolué sur tout ce qui est relatif au packaging ~~a~~ qu’il a rendu bon nombre de ces tutoriels caduque

> Il existe en ligne une pléthore de tutoriels[^4] [^5] [^6] sur comment structurer son projet et comment le déployer. Cependant l'éco-système de Python a tellement évolué depuis 2018 sur tout ce qui est relatif au *packaging* a qu'il a rendu bon nombre de ces tutoriels caduque[^7] [^8].

Reviewer: ```suggestion
Il existe en ligne une pléthore de tutoriels[^4] [^5] [^6] sur comment structurer son projet et comment le déployer. Cependant l'éco-système de Python a tellement évolué depuis 2018 sur tout ce qui est relatif au *packaging* qu'il a rendu bon nombre de ces tutoriels caduc[^7] [^8].
```

> Cet article a donc pour objectif de guider les lecteurs vers une structure de projet adaptée aux usages de 2023. Parmi les usages communs, nous avons retenu pour cet article: comment [tester son application] et comment [partager son programme à autrui].

Reviewer: ```suggestion
Cet article a donc pour objectif de guider les lecteurs vers une structure de projet adaptée aux usages de 2023. Parmi les usages communs, nous avons retenu pour cet article: comment [tester son application] et comment [partager son programme avec autrui].
```

Cet article s'appuie sur les recommandations de la [PyPA] (*Python Packaging Autority*) qui est en charge de *spécifier*, *maintenir* et *documenter* toute une série de pratiques et d'outils relatifs au partage des applications et des bibliothèques développées en Python. À défaut de recommandations de la PyPA, nous utilisons les pratiques et les outils les plus répandus. Tous les partis pris part ce document sont discutés dans le [dernier chapitre].

[tester son application]: #ajout-de-tests-automatiques
[partager son programme à autrui]: #création-dune-archive-installable
[PyPA]: https://www.pypa.io/en/latest/

> [dernier chapitre]: #choix-des-technologies

Reviewer: intéressante cette idée de séparer les liens du paragraphe, je connaissais pas

Mise en place du squelette du projet

> ------------------------------------

Reviewer: ```suggestion
## Mise en place du squelette du projet
```

*Vous avez toujours besoin de cette étape.*

Commencez par créer un dossier nommé selon le nom de votre projet, dans cet exemple nous allons nommer ce dossier `tartempion`. Déplacez vous ensuite à l'intérieur de ce dossier et créez les fichiers suivants:

* `README.md`
* `LICENSE.txt`

Le fichier `README.md` est le fichier qui décrit votre logiciel, comment l'installer, comment l'utiliser et pointe des ressources pour aller plus loin. À minima, ce fichier reprend le nom du projet ainsi qu'une courte description:

    tartempion
    ==========

    Un projet vide juste pour montrer comment structurer un projet en python.

Le fichier `LICENSE.txt` est un fichier à vocation légale, il explique qui a le droit de modifier et/ou de partager votre projet. Par défaut, en l'absence d'un tel fichier, vous êtes protégé par le droit d'auteur, vous êtes le seul détenteur du logiciel, seul vous avez le droit de le modifier, de l'utiliser et de le commercialiser. Pour qu'il n'y ait aucune ambiguïté, vous pouvez rappeler votre droit d'auteur en écrivant ceci dans le fichier:

>     Copyright (c) votre nom et prénom

Reviewer: Pas tout à fait d'accord avec ce passage, les notions de droit d'auteur et de copyright sont différentes, de plus, le copyright n'est techniquement pas reconnu en France (ni en Belgique).

Author: Tu veux bien proposer un texte alternatif ?

> Si vous voulez plutôt rendre votre projet open-source, avec plus ou moins de conditions, vous pouvez copier/coller le texte d'une licence existante. Vous trouverez une liste de licence communes sur le site https://choosealicense.com/.

Reviewer: Note: l'open source ne retire pas la paternité d'un projet, ni le droit droit d'auteur (en France), ce passage donne l'impression que le droit d'auteur serait opposé à l'open source

Author: Tu veux bien proposer un texte alternatif ?

Votre projet devrait ressembler à ceci:

    tartempion/
      LICENSE.txt
      README.md

En étant dans le dossier de votre projet, donc en étant à l'intérieur du dossier `tartempion`, créez un second dossier du même nom, c'est à dire créez à nouveau un dossier `tartempion`. Créez ensuite dans ce nouveau dossier, les deux fichiers suivants:

* `__init__.py`
* `__main__.py`

> Ce second dossier contiendra votre programme, c'est-à-dire tous vos fichiers python (à l'exception des tests automatiques). Pour différencier les deux dossiers, nous appellerons le 1e dossier (celui qui contient le fichier `README.md`) la *racine du projet* (*project root*).

Reviewer: ```suggestion
Ce second dossier contiendra votre programme, c'est-à-dire tous vos fichiers python (à l'exception des tests automatiques). Pour différencier les deux dossiers, nous appellerons le premier dossier (celui qui contient le fichier `README.md`) la *racine du projet* (*project root*).
```
ou `1er`, mais pas `1e`

> Le fichier `__init__.py` fait en sorte que le dossier soit reconnu comme un module par python et sera automatiquement exécuté à chaque fois que le projet sera importé et/ou exécuté.

Reviewer: ```suggestion
Le fichier `__init__.py` permet à python de reconnaitre le dossier comme un module et sera automatiquement exécuté à chaque fois que le projet sera importé et/ou exécuté.
```
j'aime pas les "fait en sorte de..." et autres paraphrases

> Le fichier `__main__.py` est le point d'entré de votre application, il contient le code qui sera exécuté (après `__init__.py`) lorsque vous allez lancer votre programme. Ce fichier n'est pas nécessaire si vous compter développer une bibliothèque de fonctions et non pas un programme exécutable.

Reviewer 2: Si vous comptez*

> Le fichier `__main__.py` est le point d'entré de votre application, il contient le code qui sera exécuté (après `__init__.py`) lorsque vous allez lancer votre programme. Ce fichier n'est pas nécessaire si vous compter développer une bibliothèque de fonctions et non pas un programme exécutable.

Reviewer: ```suggestion
Le fichier `__main__.py` est le point d'entrée de votre application, il contient le code qui sera exécuté (après `__init__.py`) lorsque vous lancerez votre programme. Ce fichier n'est pas nécessaire si vous comptez développer une bibliothèque de fonctions et non pas un programme exécutable.
```

> Au terme de cette étape, votre projet ressemble à ceci:

Reviewer 2: J'aurais utilisé du conditionnel ici ou alors une tournure comme "doit ressembler"/"devrait ressembler".

> Au terme de cette étape, votre projet ressemble à ceci:

Reviewer: ```suggestion
Au terme de cette étape, votre projet devrait ressembler à ceci:
```

Reviewer: d'autant plus que le conditionnel est utilisé plus loin

    tartempion/
      tartempion/
        __init__.py
        __main__.py
      LICENSE.txt
      README.md

### Lancement du projet

> Pour constater que cette structure minimale fonctionne correctement, vous pouvez écrire ceci dans le fichier `__init__.py`:

Reviewer 2: J'aurai employé "vérifier" au lieu de "constater" ici

    print("Dans __init__.py")
    print("  file:", __file__)
    print("  name:", __name__)
    print("  package:", __package__)

et ceci dans le fichier `__main__.py`:

    print("Dans __main__.py")
    print("  file:", __file__)
    print("  name:", __name__)
    print("  package:", __package__)

> Ouvrez alors un terminal à la *racine du projet* (le 1e dossier `tartempion`) et exécutez la commande suivante pour constater qu'il est possible d'*importer* votre projet comme une *bibliothèque de fonctions* (sur windows, remplacez `python3` par `py`):

Reviewer: ```suggestion
Ouvrez alors un terminal à la *racine du projet* (le 1er dossier `tartempion`) et exécutez la commande suivante pour constater qu'il est possible d'*importer* votre projet comme une *bibliothèque de fonctions* (sur windows, remplacez `python3` par `py`):
```

    python3 -c "import tartempion"

La commande va ouvrir un interpréteur python, exécuter l'instruction `import tartempion` et quitter. Le résultat de cette commande devrait être:

    Dans __init__.py
      file: __init__.py
      name: tartempion
      package: tartempion

Exécutez ensuite la commande suivante pour constater qu'il est possible d'*exécuter* votre projet comme un *programme* (sur windows, remplacez `python3` par `py`):

    python3 -m tartempion

La commande va charger le module dans un interpréteur python et l'exécuter. Le résultat de cette commande devrait être:

    Dans __init__.py
      file: __init__.py
      name: tartempion
      package: tartempion
    Dans __main__.py
      file: __main__.py
      name: __main__
      package: tartempion    

> Une fois que vous avez vérifier que votre programme fonctionnait bien, vous pouvez vider les deux fichiers, c'est-à-dire retirer tous les prints. Ne supprimez pas les fichiers.

Reviewer 2: avez vérifié*

> Une fois que vous avez vérifier que votre programme fonctionnait bien, vous pouvez vider les deux fichiers, c'est-à-dire retirer tous les prints. Ne supprimez pas les fichiers.

Reviewer: ```suggestion
Une fois que vous avez vérifié que votre programme fonctionne bien, vous pouvez vider les deux fichiers, c'est-à-dire retirer tous les `print`. Ne supprimez pas les fichiers.
```

Adaptation d'un programme existant

> ----------------------------------

Reviewer: ```suggestion
## Adaptation d'un programme existant
```

> *Vous avez besoin de cette étape si vous aviez déjà commencé à faire un programme mais qu'il n'était pas correctement structuré.*

Reviewer 2: J'aurais employé du passé composé ici plutôt que du plus que parfait.

Vous avez deux dossier, un dossier mal structuré qui contient votre ancien programme et un dossier bien structuré (mais presque vide) que vous avez créé en suivant les instructions précédentes. Nous allons appeler ici ces dossiers `old_tartempion` et `new_tartempion`.

> ### Structure dans l'ancien projet

Reviewer: ```suggestion
### Structure de l'ancien projet
```

Reviewer: le projet, c'est l'ensemble des fichiers/dossiers

Votre dossier `old_tartempion` doit certainement avoir une structure similaire à la structure suivante:

    old_tartempion/
      tartempion.py
      tartes/
        __init__.py
        tarte.py
        tarte_cerise.py
        tarte_pomme.py

> D'un point de vu des imports, le contenu de vos fichiers doit certainement ressembler à quelque chose dans le genre de:

Reviewer: ```suggestion
Du point de vu des imports, le contenu de vos fichiers doit certainement ressembler à quelque chose dans ce genre:
```

```py
# tartempion.py
from tartes.tarte_cerise import TarteCerise

tarte = TarteCerise()
print(tarte)
```

```py
# __init__.py
# vide
```

```py
# tarte.py
class Tarte:
    ...
```

```py
# tarte_cerise.py
from . import tarte

class TarteCerise(tarte.Tarte):
    ...
```

```py
# tarte_pomme.py
from .tarte import Tarte

class TartePomme(Tarte):
    ...
```
    

> Vous devez certainement lancer votre programme soit en faisant `python3 tartempion.py` depuis votre terminal soit en cliquant sur le bouton "run" tout en étant sur le fichier `tartempion.py` dans votre IDE.

Reviewer: ```suggestion
Vous devrez certainement lancer votre programme soit en faisant `python3 tartempion.py` depuis votre terminal soit en cliquant sur le bouton "run" tout en étant sur le fichier `tartempion.py` dans votre IDE.
```

Author: je ne gardez pas cette suggestion, je pose un postulat ici "une application structurée de la sorte, vous la lancez de cette manière", le "devez" est comme le "lancez" ici. Tu as peut-être une autre suggestion ?

> Si la structure de votre projet actuel ne ressemble pas à celle décrite ici ou bien si vous n'êtes pas sur alors venez nous demander de l'aide sur Discord. Faites bien attention à nous expliquer comment vos fichiers sont structurés, comment vous faites vos imports et surtout comment vous lancez votre programme.

Reviewer 2: sûr* au lieu de "sur" dans "n'êtes pas sur".

> Si la structure de votre projet actuel ne ressemble pas à celle décrite ici ou bien si vous n'êtes pas sur alors venez nous demander de l'aide sur Discord. Faites bien attention à nous expliquer comment vos fichiers sont structurés, comment vous faites vos imports et surtout comment vous lancez votre programme.

Reviewer: ```suggestion
Si la structure de votre projet actuel ne ressemble pas à celle décrite ici ou bien si vous n'êtes pas sûr alors venez nous demander de l'aide sur Discord. Faites bien attention à nous expliquer comment vos fichiers sont structurés, comment vous faites vos imports et surtout comment vous lancez votre programme.
```

> ### Structure dans le nouveau projet

Reviewer: ```suggestion
### Structure du nouveau projet
```

Après adaptation, la structure de votre projet devrait plutôt ressembler à ceci:

    new_tartempion/
      tartempion/
        __init__.py
        __main__.py
        tartes/
            __init__.py
            tarte.py
            tarte_cerise.py
            tarte_pomme.py
      LICENSE.txt
      README.md

> Le fichier `tartempions.py` est devenu le fichier `__main__.py` qui se trouve à l'intérieur du module `tartempion`. Un fichier (vide) `__init__.py` a été rajoute à côté. Le module `tartes` a été déplacé à l'intérieur du module `tartempion`.

Reviewer: ```suggestion
Le fichier `tartempions.py` est devenu le fichier `__main__.py` qui se trouve à l'intérieur du module `tartempion`. Un fichier (vide) `__init__.py` a été rajouté à côté. Le module `tartes` a été déplacé à l'intérieur du module `tartempion`.
```

> Niveau imports dans vos fichiers, vous devez corriger toutes vos lignes où vous avez importés des modules de votre projet à l'aide d'un import absolu. Un import est absolu lorsque ce que vous importez ne commence par par un point. 

Reviewer: ```suggestion
Niveau imports dans vos fichiers, vous devez corriger toutes vos lignes où vous avez importé des modules de votre projet à l'aide d'un import absolu. Un import est absolu lorsque ce que vos imports ne commencent pas par un point. 
```

Exemple d'imports absolus (ne commencent pas par un point):

```py
import tartes
from tartes import tarte_cerise
import tartes.tarte_cerise
from tartes.tarte_cerise import TarteCerise
```

Exemple d'imports relatifs (commencent par un point):

```py
from . import tarte
from .tarte import TarteCerise
```

Vous pouvez en apprendre plus sur les différences entre imports absolus et imports relatifs dans [cet autre article](relatif-vs-absolu-demystifions-les-imports-python)

En reprenant la structure de `old_tartempion` en exemple ici, le seul import que vous devez corriger est celui qui provient du fichier `tartempion.py` (renommé `__main__.py` dans la nouvelle structure). Dans votre propre projet, vous devrez certainement en corriger plus.

```py
# tartempion.py
from .tartes.tarte_cerise import TarteCerise
#    ^ya un point en plus ici

tarte = TarteCerise()
print(tarte)
```

Pour lancer votre projet depuis votre terminal, vous devez maintenant faire `python3 -m tartempion` (sur Windows, remplacez `python3` par `py`) en étant dans le dossier `new_tartempion`.

Les étapes nécessaires pour configurer VSCode et PyCharm sont expliquées plus bas.

Isolation dans un environnement virtuel

> ---------------------------------------

Reviewer: ```suggestion
## Isolation dans un environnement virtuel
```

*Vous avez aussi besoin de cette étape si vous voulez installer avec pip des applications, des outils ou bibliothèques supplémentaires comme pylint, requests ou pygame.*

### Introduction

> Ouvrez un terminal et entrez la commande `python3 -m pip list` (sur windows, remplacez `python3` par `py`). Le résultat de cette commande devrait lister tous les bibliothèques supplémentaires qui sont installés pour Python sur votre machine.

Reviewer 2: bibliothèques supplémentaires qui sont installées*

> Ouvrez un terminal et entrez la commande `python3 -m pip list` (sur windows, remplacez `python3` par `py`). Le résultat de cette commande devrait lister tous les bibliothèques supplémentaires qui sont installés pour Python sur votre machine.

Reviewer: ```suggestion
Ouvrez un terminal et entrez la commande `python3 -m pip list` (sur windows, remplacez `python3` par `py`). Le résultat de cette commande devrait lister tous les bibliothèques supplémentaires qui sont installées pour Python sur votre machine.
```

Si vous êtes sur Linux ou MacOS, cette liste devrait déjà être bien remplie comme les deux systèmes utilisent Python en interne pour toute une série de routines systèmes. Si vous êtes sur Windows, cette liste devrait être vide ou presque vide.

Vous pouvez répéter la commande avec l'option `--user` pour lister les bibliothèques qui sont installés pour votre utilisateur propre, la liste devrait être (beaucoup) plus courte.

Les paquets qui sont installées au niveau du système sont les paquets que le système gère lui-même, par exemple avec `apt` sur Ubuntu, vous ne devriez jamais y toucher vous-même, ni installer de nouvelles bibliothèques, ni en supprimer. Les paquets installées au niveau de votre utilisateur propre sont des outils dont vous avez besoin dans votre vie de tous les jours, par exemple pour faire du scripting.

> Lorsque vous développez un nouveau projet, il est important d'isoler les bibliothèques utilisés pour ce projet du reste des bibliothèques de votre système. Vous ne voudriez pas écraser une bibliothèque utilisée par votre système par une version trop récente/ancienne dans votre projet et ainsi compromettre le bon fonctionnement de votre système.

Reviewer 2: utilisés -> utilisées

> Lorsque vous développez un nouveau projet, il est important d'isoler les bibliothèques utilisés pour ce projet du reste des bibliothèques de votre système. Vous ne voudriez pas écraser une bibliothèque utilisée par votre système par une version trop récente/ancienne dans votre projet et ainsi compromettre le bon fonctionnement de votre système.

Reviewer: ```suggestion
Lorsque vous développez un nouveau projet, il est important d'isoler les bibliothèques utilisées pour ce projet du reste des bibliothèques de votre système. Vous ne voudriez pas écraser une bibliothèque utilisée par votre système par une version trop récente/ancienne dans votre projet et ainsi compromettre le bon fonctionnement de votre système.
```

> En Python, pour isoler les bibliothèques utilisées d'un un endroit à autre, nous utilisons des environnement virtuel.

Reviewer 2: Il y a un "un" mal positionné vers "d'un un endroit à autre".

Il manque aussi le pluriel à "des environnement virtuel".

> En Python, pour isoler les bibliothèques utilisées d'un un endroit à autre, nous utilisons des environnement virtuel.

Reviewer: ```suggestion
En Python, pour isoler les bibliothèques utilisées d'un un endroit à autre, nous utilisons des environnements virtuels.
```

### Créer un environnement virtuel

Ouvrez un terminal à la racine de votre projet et entrez la commande `python3 -m venv .env` (remplacez `python3` par `py` sur Windows). La commande va créer un nouvel environnement virtuel dans le dossier `.env` à la racine de votre projet.

Une fois l'environnement créé, vous pouvez vérifier qu'il fonctionne bien en ouvrant l'interpréteur Python lié à ce nouvel environnement. Sur MacOS/Linux, vous pouvez exécuter la commande `./.env/bin/python` dans votre terminal, sur Windows vous pouvez faire `.\.env\Scripts\python`, dans les deux cas l'exécution de cette commande devrait ouvrir l'interpréteur que vous pouvez fermer en appelant la fonction `exit()`.

> Vous pouvez aussi lister les modules supplémentaires qui sont installés dans cet environnement virtuel en faisant `./.env/bin/pip list` (`.\.env\Scripts\` sur Windows), vous devriez constater que seuls `pip` (et peut-être `setuptools`) est installé malgré qu'il peut y avoir de nombreux autres paquets installés sur votre système.

Reviewer: ```suggestion
Vous pouvez aussi lister les modules supplémentaires qui sont installés dans cet environnement virtuel en faisant `./.env/bin/pip list` (`.\.env\Scripts\` sur Windows), vous devriez constater que seul `pip` (et peut-être aussi `setuptools`) est installé malgré la présence possible de nombreux autres paquets sur votre système.
```

Author: j'ai opté pour "possible présence"

Pour reconfigurer son terminal, ou comme nous disons dans le jargon *activer son environnement virtuel* vous pouvez *sourcer* le script `activate` qui se trouve à côté de `python` et `pip` dans le dossier `bin` de votre environnement virtuel: `source ./.env/bin/activate` (sur Windows, pas besoin de `source`, appelez `.\.env\Scripts\activate.bat` directement à la place).

Pour vérifier que la reconfiguration (l'activation) s'est bien déroulé, vous pouvez exécuter la commande `pip --version` (sans donner le chemin complet) et vérifier que le `pip` utilisé est celui de votre environnement virtuel.

    $ pip --version
    pip 22.0.2 from /tmp/.env/lib/python3.10/site-packages/pip (python 3.10)

Pour faire la reconfiguration inverse, réutiliser les python et pip du système, ou comme nous disons dans le jargon *désactiver son environnement virtuel*, vous pouvez simplement exécuter la commande `deactivate` (`deactivate.bat` sur Windows)

[venv]: https://docs.python.org/3/library/venv.html

Ajout de tests automatiques

> ---------------------------

Reviewer: ```suggestion
## Ajout de tests automatiques
```

*Vous avez besoin de cette étape lorsque vous comptez vérifier le fonctionnement de votre application à l'aide de tests automatiques*

Ouvrez un terminal à la racine de votre projet et créez un nouveau dossier `tests`, dans ce dossier créez deux fichiers: `__init__.py` et `test_tartempion.py`. La structure de votre projet devrait donc ressembler à ceci:

    tartempion/
      tartempion/
        __init__.py
        __main__.py
      tests/
        __init__.py
        test_tartempion.py
      LICENSE.txt
      README.md

> Vous pouvez laisser le fichier `__init__.py` vide, il sert uniquement à marquer que le dossier est un package pour python. À l'intérieur du fichier `test_tartempion.py` vous pouvez écrire ceci:

Reviewer: ```suggestion
Vous pouvez laisser le fichier `__init__.py` vide, il sert uniquement à indiquer que le dossier est un package pour python. À l'intérieur du fichier `test_tartempion.py` vous pouvez écrire ceci:
```

    import unittest
    import tartempion

    class TartempionTest(unittest.TestCase):
        def test_tartempion_loaded(self):
            self.assertTrue(tartempion)

> Une fois le test écrit, assurez vous de vous positionner à la racine de votre projet dans votre terminal. Vous pouvez ensuite lancer les tests avec la commande `python3 -m unittest` (utilisez `py -m unittest` sur Windows, pensez à passer par votre environnement virtuel si vous en avez un). Vous devrez avoir un résultat comme suit:

Reviewer: ```suggestion
Une fois le test écrit, assurez vous de vous positionner à la racine de votre projet dans votre terminal. Vous pouvez ensuite lancer les tests avec la commande `python3 -m unittest` (utilisez `py -m unittest` sur Windows, pensez à passer par votre environnement virtuel si vous en avez un). Vous devriez avoir un résultat comme suit:
```

    .
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s

    OK

Chaque `.` sur la première ligne représente un test qui est passé, les tests échoués apparaissent avec un `!`. Vous pouvez aussi utiliser l'option `-v` si vous voulez plutôt énumérer le nom de tests un à un:

    test_tartempion (tests.test_tartempion.TestTartempion) ... ok

    ----------------------------------------------------------------------
    Ran 1 test in 0.000s

    OK

Le fonctionnement général d'[unittest] est qu'il chargera de lui-même tous les fichiers à l'intérieur du dossier `tests` qui commencent par `test_`. À l'intérieur de ces modules python, il exécutera de lui-même toutes les méthodes qui commencent par `test_` dans les classes qui héritent de `TestCase`.

> Vous pouvez ajouter autant de modules de tests et de classe que vous voulez. Il peut être judicieux de définir les tests dans des modules différents en fonction de la partie du code qu'ils testent.

Reviewer: ```suggestion
Vous pouvez ajouter autant de modules de tests et de classes que vous voulez. Il peut être judicieux de définir les tests dans des modules différents en fonction de la partie du code qu'ils testent.
```

[unittest]: https://docs.python.org/3/library/unittest.html

Création d'une archive installable

> ----------------------------------

Reviewer: ```suggestion
## Création d'une archive installable
```

Assurez-vous ensuite que ces libs sont également installées au niveau de votre environnement virtuel (si vous en utilisez un): `.env/bin/pip install --upgrade build pip setuptools wheel` (`.env\Scripts\pip` sur Windows, ou bien activez votre environnement virtuel).

### pyproject.toml

Une fois les dépendances installées, vous pouvez définir les méta-données relatives à votre projet dans un nouveau fichier `pyproject.toml` que vous créez à la racine de votre projet. Nous vous proposons le fichier suivant à titre d'exemple:

    [project]
    name = "tartempion_NaN"
    version = "0.0.1"
    description = "Une courte description sur une ligne"
    readme = "README.md"
    license = { file = "LICENSE.txt" }
    authors = [
        { name = "Bob", email = "bob@example.com" },
    ]
    dependencies = [
        # "pygame",
        # "requests",
    ]

    [build-system]
    requires = ["setuptools>=61.0", "wheel"]
    build-backend = "setuptools.build_meta"

Le fichier utilise un format relativement nouveau, le [TOML], il s'agit d'un format de fichier adapté aux fichiers de configuration.

> La première partie du fichier, `[project]`, défini l'ensemble des méta-données relatives à votre projet. Il existe en tout plus d'une vingtaine de méta-données différentes, nous avons décider de ne garder que les plus essentielles d'entre elles ici. La liste complète est défini dans la PEP [621]. Pensez à changer les valeurs pour le nom, la description et les auteurs. Pensez aussi à déclarer vos dépendances, si vous en avez, pour qu'elles soient automatiquement installées avec votre projet.

Reviewer: ```suggestion
La première partie du fichier, `[project]`, défini l'ensemble des méta-données relatives à votre projet. Il existe en tout plus d'une vingtaine de méta-données différentes, nous avons décidé de ne garder que les plus essentielles d'entre elles ici. La liste complète est défini dans la PEP [621]. Pensez à changer les valeurs pour le nom, la description et les auteurs. Pensez aussi à déclarer vos dépendances, si vous en avez, pour qu'elles soient automatiquement installées avec votre projet.
```

> **Important**, le champ `name` dans ce fichier n'est pas obligé d'être le même que le nom de votre dossier. Il est d'ailleurs conseillé d'ajouter votre pseudo ou un autre élément unique à la fin pour vous assurer qu'il sera unique sur internet. Dans cet exemple nous avons ajouté `_NaN` pour "Not a Name", ce blog.

Reviewer: ```suggestion
**Important**, le champ `name` dans ce fichier n'est pas obligatoirement le même que le nom de votre dossier. Il est d'ailleurs conseillé d'ajouter votre pseudo ou un autre élément unique à la fin pour vous assurer qu'il sera unique sur internet. Dans cet exemple nous avons ajouté `_NaN` pour "Not a Name", ce blog.
```

> La seconde partie du fichier, `[build-system]`, décrit quels outils seront utilisées pour créer les archives pour votre projet. Cette partie est obligatoire mais vous pouvez vous contenter de garder celle de l'exemple.

Reviewer: ```suggestion
La seconde partie du fichier, `[build-system]`, décrit quels outils seront utilisés pour créer les archives pour votre projet. Cette partie est obligatoire mais vous pouvez vous contenter de garder celle de l'exemple.
```

> ### build

Reviewer: Build ?
je sais qu'il s'agit d'une commande mais un titre sans maj je trouve que ça fait chelou

> Une fois le fichier `pyproject.toml` compléter, vous pouvez utiliser [build] pour créer une archive de votre fichier. Placez-vous à la racine de votre projet et lancer la commande `.env/bin/pyproject-build` (`.env\Scripts\pyproject-build` sur Windows). La commande est très bavarde, parmi toutes les lignes affichées vous devriez voir celles-ci:

Reviewer: ```suggestion
Une fois le fichier `pyproject.toml` complété, vous pouvez utiliser [build] pour créer une archive de votre fichier. Placez-vous à la racine de votre projet et lancer la commande `.env/bin/pyproject-build` (`.env\Scripts\pyproject-build` sur Windows). La commande est très verbeuse, parmi toutes les lignes affichées vous devriez voir celles-ci:
```

    * Creating venv isolated environment...
    * Installing packages in isolated environment... (setuptools)
    * Getting build dependencies for sdist...
    * Building sdist...
    * Building wheel from sdist
    * Creating venv isolated environment...
    * Installing packages in isolated environment... (setuptools)
    * Getting build dependencies for wheel...
    * Installing packages in isolated environment... (wheel)
    * Building wheel...
    Successfully built tartempion-0.0.1.tar.gz and tartempion-0.0.1-py3-none-any.whl

Le programme aura créé plusieurs fichiers et dossiers à la racine de votre projet, parmi ces dossiers un dossier `dist/` qui contient les deux archives créées: `tartempion-0.0.1.tar.gz` (`.zip` sur Windows) et `tartempion-0.0.1-py3-none-any.whl`.

> La première archive au format `.tar.gz` ou `.zip` est une *distribution source*, elle est principalement à destination des développeurs. La seconde archive au format `.whl` est une *distribution wheel* (sur roulette), elle est principalement à destination des utilisateurs.

Reviewer: rofl pour la traduction de wheel x)

Author: ahah pun intended :D

Chacune de ces deux archives est installable via [pip]. Vous pouvez donc les partagez autours de vous (préférez partager la wheel) et vos utilisateurs pourront l'installer sur leur propre machine.

[TOML]: https://toml.io/fr/
[621]:https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#declaring-project-metadata
[build]: https://pypa-build.readthedocs.io/en/stable/index.html
[setuptools]: https://setuptools.pypa.io/en/latest/
[wheel]: https://wheel.readthedocs.io/en/latest/

Configuration de VSCode

> -----------------------

Reviewer: ```suggestion
## Configuration de VSCode
```

> *Vous avez besoin de cette étape si vous voulez que VSCode arrête de lever des erreurs pour des problèmes d'import et/ou pouvoir lancer votre projet directement depuis la console de VSCode.*

Reviewer: je suis pas certain que ça soit pertinent de parler de vscode/pycharm dans cet article, la configuration de son éditeur/IDE me semble être un sujet à part, mais bon

Author: ce sont des questions vraiment récurrentes, typiquement le gars va installer une lib en global mais sans l'installer dans son environnement virtuel... il vient ensuite pleurer "pk il trouv pa la lib ? jlai instalé avec pip pourtant"

Installez l'extension Python de Microsoft pour VSCode, cette extension ajoute le nécessaire pour exécuter et debugger du code python directement depuis VSCode.

Ouvrez une nouvelle fenêtre dans VSCode (File > New Window), ouvrez le dossier qui contient votre projet. Une fois votre projet ouvert, assurez-vous que votre environnement virtuel a été chargé en regardant le coin inférieur droit, vérifiez de bien lire `('.env': venv)` précédé de la version de Python avec laquelle vous aviez créé votre projet.

Pour lancer votre module directement depuis VSCode et profiter du debugger intégrer, ajoutez une nouvelle configuration pour le lancement de votre programme (Run > Add Configuration). VSCode vous proposera une configuration parmi d'autres, sélectionnez "**Module** Debug a Python module by invoking it '-m'", il devrait être le second dans la liste. VSCode vous demande ensuite de donner le nom de votre module, nous entrons `tartempion` dans notre cas, vous inscrivez le nom de votre module. Une fois la nouvelle configuration ajoutée, il vous reste à lancer le debugger avec F5 (ou bien Run > Start Debugging).

Configuration de PyCharm

> ------------------------

Reviewer: ```suggestion
## Configuration de PyCharm
```

*Vous avez besoin de cette étape si vous voulez que PyCharm arrête de lever des erreurs pour des problèmes d'import et/ou pouvoir lancer votre projet directement depuis la console de PyCharm.*

> <!-- Lazor: arrêtez de casser les couilles et utilisez Sublime Text avec un terminal. -->

Reviewer: subtile

Author: bill

Author: en vrai j'attend que quelqu'un propose une PR pour la partie pycharm, je ne compte pas l'écrire moi-même

Choix des technologies

> ----------------------

Reviewer: ```suggestion
## Choix des technologies
```

> Tout au long de ce document, il n'a jamais été présenté qu'une seule approche, un seul choix. Il s'agit d'un choix assumé par la nécessité de dissimuler aux débutants toute la complexité réelle qui se cache derrière le *packaging* en Python. Dans cette section, nous présentons les alternatives connues pour chacune de nos décisions et nous justifions nos choix.

Reviewer: très bonne idée cette section

### Squelette du projet

Il existe deux manières de structurer un projet python[^9], ce document a présenté la structure dite "à plat" (*flat-layout*) mais certains projets préfèrent la structure dite "imbriquée" (*src-layout* ou *nested-layout*).

Structure à plat:

    tartempion/
      tartempion/
        __init__.py
        __main__.py
      tests/
        test_tartempion.py
      README.md

Structure imbriquée:

    tartempion/
      src/
        tartempion/
          __init__.py
          __main__.py
      tests/
        test_tartempion.py
      README.md

> Nous avons préféré présenter la structure à plat car elle est plus simple. La structure imbriquée n'est utile que lorsqu'un même projet se décline en plusieurs modules co-dépendants les uns des autres qui ne doivent pas partager un même *namespace*. Plutôt rare comme cas de figure.

Reviewer: ```suggestion
Nous avons préféré présenter la structure à plat car elle est plus simple. La structure imbriquée n'est utile que lorsqu'un même projet se décline en plusieurs modules dépendants les uns des autres qui ne doivent pas partager un même *namespace*. Plutôt rare comme cas de figure.
```
"co-dependants" ou "dependants les uns des autres", il faut choisir

> Un autre avantage de la structure à plat est de pouvoir lancer le programme et les tests automatiques depuis la racine du projet sans installation. Avec la structure imbriquée, il est obligatoire de recourir à un paquet installé en mode éditable dans l'environnement virtuel courant pour lancer et tester l'application. Cependant cet avantage peut aussi s'avéré désavantageux lorsqu'il s'agit de créer un package, comme certains fichiers *pourraient* apparaître comme accessible sans réellement l'être[^10].

Reviewer: ```suggestion
Un autre avantage de la structure à plat est de pouvoir lancer le programme et les tests automatiques depuis la racine du projet sans installation. Avec la structure imbriquée, il est obligatoire de recourir à un paquet installé en mode éditable dans l'environnement virtuel courant pour lancer et tester l'application. Cependant cet avantage peut aussi s'avérer désavantageux lorsqu'il s'agit de créer un package, par exemple certains fichiers *pourraient* apparaître comme accessibles sans réellement l'être[^10].
```
un avantage désavantageux, hmm

Author: j'ai reformulé

### Format du README

Il existe plusieurs formats de fichier assez répandus pour la rédaction des fichiers README. Les plus communs sont le format texte, le format Markdown et le format reStructuredText. Sur <pypi.org>, le format utilisé par défaut est reStructuredText, cependant lorsqu'un projet est créé depuis [Github], le fichier README sera au format Markdown.

Il y a une grosse tendance à favoriser le Markdown pour l'écriture des fichiers README et cette tendance n'est pas spécifique à Python. Puisque le Markdown est également supporté sur <pypi.org> (via la directive `long_description_content_type=text/markdown`) et que le Markdown est généralement considéré comme plus simple à utiliser que le reStructuredText, nous avons choisi le format Markdown pour le README.

[Github]: https://github.com

### Environnement Virtuel

> Il existe plusieurs logiciels capable de créer et de gérer des environnements virtuels parmi lesquels [venv], [virtualenv] et [pyenv-virtualenv]. Il existe aussi plusieurs logiciels de gestion de projet python "tout-en-un" qui intègrent un environnement virtuel par défaut comme [hatch], [pdm] ou [poetry].

Reviewer: ```suggestion
Il existe plusieurs logiciels capables de créer et de gérer des environnements virtuels parmi lesquels [venv], [virtualenv] et [pyenv-virtualenv]. Il existe aussi plusieurs logiciels de gestion de projet python "tout-en-un" qui intègrent un environnement virtuel par défaut comme [hatch], [pdm] ou [poetry].
```

Parmi tous ces outil, [venv] est disponible de base et était recommandé par la PyPA sur sa page des outils recommandés[^11] au moment de la rédaction de ce document.

[venv]: https://docs.python.org/3/library/venv.html
[virtualenv]: https://virtualenv.pypa.io/en/stable/index.html
[pyenv-virtualenv]: https://github.com/pyenv/pyenv-virtualenv
[hatch]: https://hatch.pypa.io/latest/
[pdm]: https://pdm.fming.dev/
[poetry]: https://python-poetry.org/
[setuptools]: https://setuptools.readthedocs.io/en/latest/

### Tests automatiques

> Il existe plusieurs bibliothèques pour écrire et lancer des tests unitaires sur Python. On peut citer [unittest], [pytest], [nose], [ward] et [hammett]. Parmi tous ces bibliothèques, les deux plus répandues sont [unittest] et [pytest]. Malgré que de nombreuses personnes préfèrent [pytest] à [unittest], [unittest] reste utilisé par une majorité de projets et est disponible de base.

Reviewer: ```suggestion
Il existe plusieurs bibliothèques pour écrire et lancer des tests unitaires sur Python. On peut citer [unittest], [pytest], [nose], [ward] et [hammett]. Parmi toutes ces bibliothèques, les deux plus répandues sont [unittest] et [pytest]. Malgré une plus grande préférence des gens pour [pytest] à [unittest], [unittest] reste utilisé par une majorité de projets et est disponible de base.
```
"malgré que" = bad

Author: j'ai reformulé

[unittest]: https://docs.python.org/3/library/unittest.html
[pytest]: https://docs.pytest.org
[nose]: https://nose.readthedocs.io/en/latest/
[ward]: https://ward.readthedocs.io/en/latest/
[hammett]: https://github.com/boxed/hammett

### Distribution

Depuis 2018, de nombreux systèmes ont vu le jour avec comme objectif de simplifier la gestion des projets en Python. On peut citer [poetry] avec son système de gestion de dépendances novateur, [pdm] et sa volonté de fournir un outil semblable à npm, [flit] et sa volonté de fournir un logiciel *simple* et pour finir [hatch] qui se veut être un outil tout-en-un et qui est de plus en plus répandu.

> Il a existé une époque entre 2018 et 2021 pendant laquelle les mises à jour de [setuptools] lagaient derrière ces nouveaux outils en terme de l'adoption des standards (PEP-[517]/[518]/[621]) et pendant laquelle il nous aurait été difficile de suivre la PyPA quant ses recommandations. Heureusement, la [version 61] de [setuptools] (avril 2021) a rajouté le support pour la PEP-[621] (*Storing project metadata in pyproject.toml*) et il est depuis lors raisonnable d'utiliser setuptools à nouveau.

Reviewer: ```suggestion
Il a existé une époque entre 2018 et 2021 pendant laquelle les mises à jour de [setuptools] stagnaient derrière ces nouveaux outils en terme d'adoption des standards (PEP-[517]/[518]/[621]) et pendant laquelle il nous aurait été difficile de suivre la PyPA dans ses recommandations. Heureusement, la [version 61] de [setuptools] (avril 2021) a ajouté le support pour la PEP-[621] (*Storing project metadata in pyproject.toml*) et il est depuis lors raisonnable d'utiliser setuptools à nouveau.
```

> Nous avons utilisé setuptools dans cet article comme il est compatible avec les PEP [517] et [621], qu'il semble toujours être le seul outil capable de gérer des extensions écrites en C et qu'il est toujours le seul outil recommandé[^11] par la PyPA.

Reviewer: ```suggestion
Nous avons utilisé setuptools dans cet article car il est compatible avec les PEP [517] [621], il semble toujours être le seul outil capable de gérer des extensions écrites en C et il est toujours le seul outil recommandé[^11] par la PyPA.
```

Author: j'ai répété *car* pour chaque élément

[flit]: https://flit.readthedocs.io/en/latest/
[hatch]: https://hatch.pypa.io/latest/
[pdm]: https://pdm.fming.dev/
[poetry]: https://python-poetry.org/
[setuptools]: https://setuptools.readthedocs.io/en/latest/
[version 61]: https://setuptools.pypa.io/en/latest/history.html#v61-0-0

[517]: https://peps.python.org/pep-0517/
[518]:https://peps.python.org/pep-0518/
[621]:https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#declaring-project-metadata

### IDE

> Parmi tous les éditeurs de texte et IDE disponibles pour Python, nous n'avons décidé de n'en documenter que deux, [VSCode] et [PyCharm]. Cette décision repose sur l'expérience des auteurs quant à la fréquence des questions relative à l'utilisation combinées des environnements virtuels et des éditeurs de texte/IDE, expérience qui a été confirmée par plusieurs sondages.

Reviewer: ```suggestion
Parmi tous les éditeurs de texte et IDE disponibles pour Python, nous avons décidé de n'en documenter que deux, [VSCode] et [PyCharm]. Cette décision repose sur l'expérience des auteurs quant à la fréquence des questions relatives à l'utilisation combinées des environnements virtuels et des éditeurs de texte/IDE, expérience qui a été confirmée par plusieurs sondages.
```

Les débutants qui utilisent [VSCode] et [PyCharm] utilisent souvent le bouton "run" de leur IDE pour exécuter leur programme tandis que les débutants qui utilisent d'autres éditeurs (comme [Sublime Text]) semblent préférer passer par un terminal. Ceci s'ajoute au fait que les utilisateurs débutants de VSCode et PyCharm utilisent souvent un linter intégré à leur IDE contrairement aux utilisateurs débutants qui utilisent d'autre éditeurs.

> Ces deux aspects, le bouton "run" et le linter, font qu'il est crucial de correctement configuré ces IDE pour qu'il utilise le bon environnement virtuel. Ne pas configurer d'environnement virtuel est une source commune d'erreur chez les débutants qui se plaignent soit qu'ils n'arrivent pas à installer de bibliothèques soit que leur IDE ne reconnaît pas la bibliothèque qu'ils viennent d'installer. Les deux affirmations sont en fait la source du même problème, la bibliothèque n'a pas été installée dans le même environnement virtuel que celui utilisé dans l'IDE.

Reviewer: ```suggestion
Ces deux aspects, le bouton "run" et le linter, font qu'il est crucial de correctement configurer ces IDE pour qu'ils utilisent le bon environnement virtuel. Ne pas configurer d'environnement virtuel est une source commune d'erreur chez les débutants qui se plaignent soit qu'ils n'arrivent pas à installer de bibliothèques soit que leur IDE ne reconnaît pas la bibliothèque qu'ils viennent d'installer. Les deux affirmations sont en fait la conséquence du même problème, la bibliothèque n'a pas été installée dans le même environnement virtuel que celui utilisé dans l'IDE.
```

> Il est rare de voir des questions de ce genre pour d'autres IDE, nous n'avons donc pas chercher à les documenter dans cet article. Nous ne les recommandons d'ailleurs pas plus que d'autres éditeur de texte / IDE.

Reviewer: ```suggestion
Il est rare de voir des questions de ce genre pour d'autres IDE, nous n'avons donc pas cherché à les documenter dans cet article. Nous ne les recommandons d'ailleurs pas plus que d'autres éditeur de texte / IDE.
```
